Add option to force join order for segment queries

bx80 opened the pull request:
### Description:

Fixes #21635

This PR adds two new config options:
```
[database]
enable_segment_first_table_join_prefix = 1
enable_first_table_join_prefix = 1
```
When enabled the first option will cause a query hint to be added to segment selection queries which will suggest the join order of the first table and prevent the MySQL query planner from choosing a different starting table for the join plan.

The second option does the same but for all queries built using [LogAggregator::GenerateQuery()](https://github.com/matomo-org/matomo/blob/7a4217ec6975ff9765b38b330fdbc1ab77a55241/core/DataAccess/LogAggregator.php#L345) 

Before:
```
SELECT 
    DISTINCT log_visit.idvisit AS idvisit
FROM log_visit AS log_visit
LEFT JOIN log_link_visit_action llva ON llva.idvisit = log_visit.idvisit
LEFT JOIN log_action la ON la.idaction = llva.idaction_url
-- etc
```
After:
```
SELECT /*+ JOIN_PREFIX(log_visit) */
    DISTINCT log_visit.idvisit AS idvisit
FROM log_visit AS log_visit
LEFT JOIN log_link_visit_action llva ON llva.idvisit = log_visit.idvisit
LEFT JOIN log_action la ON la.idaction = llva.idaction_url
-- etc
```

Notes:
- ~~This only applies to queries used to build segment temporary tables, it is not currently applied to any other queries.~~
- Only the first table order is specified, the optimizer may choose any order for the other tables.
- The config options default to disabled so this will have no affect unless enabled.
- The hint should be ignored by database engines that do not support it, but it's not recommended to enable it unless running MySQL 8.0+.
- A new DatabaseConfig class has been added to easily retrieve [database] config values.
- The `queryVisitsByDimension()` and `createSegmentTable()` methods have both had their sql generation code moved to sub-methods to allow easier testing of the query sql.
- Various tests have been added to check the addition of query hints based on config values.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
Add join_prefix query hint to segment select queries

caddoo reviewed (approved):
Seems fine, I'm a bit nervous about there being no unit tests though for `addJoinPrefixHintToQuery`

Because this will be a opt-in piece of code maybe you could just add one test for this part:
```
$generalConfig = Config::getInstance()->General;
        if (empty($generalConfig['enable_segment_first_table_join_prefix']) || $generalConfig['enable_segment_first_table_join_prefix'] != "1") {
            return $sql;
        }
```
To make sure that the opt-in part is at least working

michalkleiner commented on config/global.ini.php:
```diff
@@ -877,6 +877,10 @@
 ; so it can be disabled here if necessary.
 disable_tracking_matomo_app_links = 0
 
+; Force the order of the first table when building segment queries in MySQL. This can be used to override sub-optimal
```
I wouldn't use the word 'force' here as it 'only' adds a hint, I'd suspect the engine doesn't need to respect it and it would also only work when supported, so perhaps something like..

```suggestion
; Add query hint for the order of joined tables when building segment queries in MySQL. This can be used to override sub-optimal
```

bx80 replied:
Good point, I've now added another config option but will avoid the use of force for both :+1:

michalkleiner commented on core/DbHelper.php:
```diff
@@ -352,6 +352,32 @@ public static function addOriginHintToQuery(string $sql, string $origin, ?Date $
+     */
+    public static function addJoinPrefixHintToQuery(string $sql, string $prefix): string
+    {
+        $generalConfig = Config::getInstance()->General;
+        if (empty($generalConfig['enable_segment_first_table_join_prefix']) || $generalConfig['enable_segment_first_table_join_prefix'] != "1") {
+            return $sql;
+        }
+
+        $select = 'SELECT';
+        if (0 === strpos(trim($sql), $select)) {
+            $sql = trim($sql);
+            $sql = 'SELECT /*+ JOIN_PREFIX(' . $prefix . ') */' . substr($sql, strlen($select));
+        }
+
+        return $sql;
```
Not the most pressing thing here, just a code shortening/simplification.

We can either do an early return like we have here or when a thing is dependent on a config being set and enabled, this condition check is shorter and covers both the 'is set' and 'is truthy' as the null coalesce handles the array key not set as well, and we can drop the variable.

```suggestion
        if (Config::getInstance()->General['enable_segment_first_table_join_prefix'] ?? false) {
            $select = 'SELECT';
            if (0 === strpos(trim($sql), $select)) {
                $sql = trim($sql);
                $sql = 'SELECT /*+ JOIN_PREFIX(' . $prefix . ') */' . substr($sql, strlen($select));
            }
        }

        return $sql;
```

bx80 replied:
The config check has now been moved elsewhere, but thanks for the cleaner code suggestion :+1:

michalkleiner commented:
Code change looks good, just a question — would the hint use the actual table name or the aliased name (if there's an alias used)?

bx80 commented:
The hint uses the alias if the table has one:
> If a table has an alias, hints must refer to the alias, not the table name. 

It's a bit confusing here because the table alias we use is the same as the table name!

Commit:
Added additional option to add query join hint to all generated queries, added tests

bx80 commented:
I've updated the PR: 
 - Added a second config option to apply the query join hint to all generated queries
 - Moved the segment temporary table sql generation code into a separate method to make it easier to test
 - Added three new tests to the LogAggregatorTest:
   -  Test that the join hint is applied to segment table select queries if the config is enabled
   -  Test that the join hint is NOT applied to segment table select queries if the config is disabled (we don't have any other tests for the segment table select so this seems necessary to ensure the hint isn't always being applied)
   - Test that the join hint is applied to a non-segment table generated query if the config is enabled

sgiehl commented on core/DataAccess/LogAggregator.php:
```diff
@@ -385,6 +385,12 @@ public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit
             $query['sql'] = DbHelper::addOriginHintToQuery($query['sql'], $this->queryOriginHint, $this->dateStart, $this->dateEnd, $this->sites, $this->segment);
+
+            $generalConfig = Config::getInstance()->General;
+            if (!empty($generalConfig['enable_first_table_join_prefix']) && $generalConfig['enable_first_table_join_prefix'] == "1")
```
Could be simplified by using `GeneralConfig::getConfigValue('enable_first_table_join_prefix')`

bx80 replied:
Now using the new `DatabaseConfig::getConfigValue()` method :+1:

sgiehl commented on core/DataAccess/LogAggregator.php:
```diff
@@ -401,20 +407,41 @@ public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit
-        $this->createTemporaryTable($segmentTable, $segmentSql['sql'], $segmentSql['bind']);
+        if (is_array($segmentSql) && array_key_exists('sql', $segmentSql)) {
+            $generalConfig = Config::getInstance()->General;
+            if (!empty($generalConfig['enable_segment_first_table_join_prefix']) && $generalConfig['enable_segment_first_table_join_prefix'] == "1") {
```
Can be simplified using `GeneralConfig` as well.

bx80 replied:
As above.

sgiehl commented on tests/PHPUnit/Integration/DataAccess/LogAggregatorTest.php:
```diff
@@ -341,6 +341,71 @@ public function test_getSegmentTmpTableNameWithLongPrefix()
         $this->assertEquals('logtmpsegmentcc2efa0acbd5f209', $this->logAggregator->getSegmentTmpTableName());
```
Could maybe be worth to add a test that uses the new join feature as well as max execution time hint.

bx80 replied:
I've moved the query generation code from `queryVisitsByDimension` to a new sub-method `getQueryByDimensionSql` to make it easier to test the generated sql. I've then added a new test to check the expected visits by dimension query with an origin hint, join prefix hint and a max execution time hint.

sgiehl commented on config/global.ini.php:
```diff
@@ -877,6 +877,13 @@
+
+; Add a query hint for the order of the first table for all log table queries in MySQL.
+enable_first_table_join_prefix = 0
+
```
Was wondering if this would make more sense to be part of the DB or DbReader config. As it's more related to database than general config. But don't have a strong opinion here

bx80 replied:
Yes, that probably makes the most sense :+1:
I've moved the settings to the [database] section, as part of this I've made the `GeneralConfig` class a bit easier to extend and added a new descendant class `DatabaseConfig` so we can easily get settings from the `[database]` section in the same way as we already do for `[General]`

sgiehl reviewed (changes requested):
Left some comments for possible improvements. Besides that the changes look good to me 👍

Commit:
Reworked GeneralConfig for easier inheritance, added new descendant DatabaseConfig class to retrieve [database] settings

Commit:
Move join query hint config settings to [database] section, adjusted queryVisitByDimension to allow sql testing, added new test

Commit:
Fix for different ranking query return type

bx80 commented:
@michalkleiner Yeah, I thought about doing that but was trying to limit the rework scope, in retrospect it should be quick to do and a better starting point for any other new descendants so I'll make the change now :+1:

Commit:
Further abstraction of GeneralConfig, fixed test config setting

bx80 commented:
@michalkleiner @caddoo Thanks for all the feedback, this should be ready for a final review now :smiley:

michalkleiner commented on tests/PHPUnit/Integration/DataAccess/LogAggregatorTest.php:
```diff
@@ -364,6 +430,72 @@ public function test_generateQuery_WithQueryHint_ShouldAddQueryHintAsComment()
 
+    public function test_generateQuery_ShouldAddJoinQueryHintAsCommentIfEnabled()
+    {
+        DatabaseConfig::setConfigValue('enable_first_table_join_prefix','1');
```
```suggestion
        DatabaseConfig::setConfigValue('enable_first_table_join_prefix', '1');
```

michalkleiner commented on tests/PHPUnit/Integration/DataAccess/LogAggregatorTest.php:
```diff
@@ -341,6 +342,71 @@ public function test_getSegmentTmpTableNameWithLongPrefix()
+    public function test_getSegmentTableSql_ShouldNotAddJoinHintAsCommentIfDisabled()
+    {
+        DatabaseConfig::setConfigValue('enable_segment_first_table_join_prefix', '0');
+        $query = $this->getSegmentSql();
+
+        $expected = [
```
```suggestion

        $query = $this->getSegmentSql();
        $expected = [
```

michalkleiner commented on config/global.ini.php:
```diff
@@ -49,6 +49,13 @@
 ;
 ;ignore_error_codes[] = 1105
 
+; Add a query hint for the order of joined tables when building segment queries in MySQL. This can be used to override
+; sub-optimal choices by the MYSQL optimizer and always ensure the query plan starts with the first table in the query.
+enable_segment_first_table_join_prefix = 0
+
+; Add a query hint for the order of the first table for all log table queries in MySQL.
+enable_first_table_join_prefix = 0
```
```suggestion
; Add a query hint for the order of joined tables when building segment queries in MySQL. This can be used to override
; sub-optimal choices by the MySQL optimizer and always ensure the query plan starts with the first table in the query.
enable_segment_first_table_join_prefix = 0

; Add a query hint for the order of joined tables for all log table queries in MySQL.
enable_first_table_join_prefix = 0
```

michalkleiner reviewed (approved):
Just a minor formatting/text tweak that can be skipped for now.
Code changes look good, newly added tests are passing, although I haven't run the code locally to see the actual queries.

bx80 committed:
Update tests/PHPUnit/Integration/DataAccess/LogAggregatorTest.php

Co-authored-by: Michal Kleiner <michal@innocraft.com>

bx80 committed:
Update tests/PHPUnit/Integration/DataAccess/LogAggregatorTest.php

Co-authored-by: Michal Kleiner <michal@innocraft.com>

bx80 committed:
Update config/global.ini.php

Co-authored-by: Michal Kleiner <michal@innocraft.com>

bx80 committed:
Merge branch '5.x-dev' into m21635-segment-query-join-prefix

tsteur commented:
@matomo-org/core-reviewers just seeing some slow query and wonder if the join_prefix would be also expected to be set in the subqueries? I don't think it would have been useful in that particular case, and unsure how easy it is to add. Something to consider potentially as MySQL might still make suboptimal choices maybe in subqueries? Unsure though how this may affect dependent vs independent subqueries.

<img width="1149" alt="image" src="https://github.com/matomo-org/matomo/assets/273120/590e744c-e01e-41f6-b6c3-ba52982874b8">
